File: src/shared/auth/crypto.ts
import bcrypt from "bcryptjs";
import crypto from "crypto";

const BCRYPT_ROUNDS = 12;

export const hashPassword = async (rawPassword: string): Promise<string> => {
  return bcrypt.hash(rawPassword, BCRYPT_ROUNDS);
};

export const comparePassword = async (rawPassword: string, passwordHash: string): Promise<boolean> => {
  return bcrypt.compare(rawPassword, passwordHash);
};

export const randomToken = (bytes = 32): string => {
  return crypto.randomBytes(bytes).toString("hex");
};

export const sha256 = (value: string): string => {
  return crypto.createHash("sha256").update(value).digest("hex");
};

export const slugify = (value: string): string => {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 48);
};

File: backend-go/internal/services/crypto.go
package services

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"

	"golang.org/x/crypto/bcrypt"
)

const bcryptRounds = 12

func HashPassword(raw string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(raw), bcryptRounds)
	if err != nil {
		return "", fmt.Errorf("hash password: %w", err)
	}
	return string(bytes), nil
}

func ComparePassword(raw, hash string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(raw))
	return err == nil
}

func RandomToken(bytesLen int) string {
	buf := make([]byte, bytesLen)
	if _, err := rand.Read(buf); err != nil {
		// Fallback to hex-encoded random (less secure but shouldn't happen)
		return fmt.Sprintf("%x", buf)
	}
	return hex.EncodeToString(buf)
}

func SHA256(value string) string {
	h := sha256.Sum256([]byte(value))
	return hex.EncodeToString(h[:])
}

func Slugify(value string) string {
	// Convert to lowercase
	slug := strings.ToLower(strings.TrimSpace(value))
	// Replace non-alphanumeric with hyphens
	re := regexp.MustCompile(`[^a-z0-9]+`)
	slug = re.ReplaceAllString(slug, "-")
	// Trim leading/trailing hyphens
	slug = strings.Trim(slug, "-")
	// Limit length
	if len(slug) > 48 {
		slug = slug[:48]
	}
	return slug
}
